src/GameLogic/Entities/Iceblock.ts

```diff
@@ -0,0 +1,110 @@
+import {Entity} from '../Entity';
+import {EntityType, FloorType} from '../Enums';
+import {Direction8, Direction8Utils} from '../Enums/Direction8';
+import {Level} from '../Level';
+import {Fireball} from './Fireball';
+
+export class Iceblock implements Entity {
+	public readonly type: EntityType;
+
+	public x: number;
+
+	public y: number;
+
+	public direction: Direction8;
+
+	public contains: Entity;
+
+	public melting: boolean;
+
+	public justPushed: boolean;
+
+	constructor(contains: Entity) {
+		this.type = EntityType.Iceblock;
+		this.x = 0;
+		this.y = 0;
+		this.direction = Direction8.None;
+		this.contains = contains;
+		this.melting = contains.type == EntityType.Fireball;
+		this.justPushed = false;
+	}
+
+	public update(level: Level): void {
+		if (this.justPushed) {
+			this.justPushed = false; //Don't update because we've just been moved via push()
+		} else if (this.isMoveAllowed(level, this.direction) && !this.melting) {
+			this.x += Direction8Utils.getX(this.direction);
+			this.y += Direction8Utils.getY(this.direction);
+			this.postMoveProcessing(level);
+		} else {
+			this.direction = Direction8.None;
+		}
+		if (this.melting) {
+			this.contains.x = this.x;
+			this.contains.y = this.y;
+			level.entities.addEntity(this.contains);
+			level.entities.removeEntity(this);
+		}
+	}
+
+	public clone(): Iceblock {
+		const clone = new Iceblock(this.contains);
```

Comment:
You need to also clone the contained entity. Otherwise whatever the iceblock contains can change if the original iceblock melts and the entity starts moving again.

Reply:
Yes, spotted this independently of this comment. Have fixed.